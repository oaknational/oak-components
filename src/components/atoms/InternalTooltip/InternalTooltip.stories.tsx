import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakBox } from "../OakBox";

import { InternalTooltip } from "./InternalTooltip";

import { colorArgTypes } from "@/storybook-helpers/colorStyleHelpers";
import { spacingArgTypes } from "@/storybook-helpers/spacingStyleHelpers";
import { borderArgTypes } from "@/storybook-helpers/borderStyleHelpers";
import { typographyArgTypes } from "@/storybook-helpers/typographyStyleHelpers";

const meta: Meta<typeof InternalTooltip> = {
  component: InternalTooltip,
  tags: ["autodocs"],
  title: "components/atoms/InternalTooltip",
  argTypes: {
    children: {
      control: {
        type: "text",
      },
    },
    ...colorArgTypes,
    ...spacingArgTypes,
    ...borderArgTypes,
    ...typographyArgTypes,
  },
  parameters: {
    controls: {
      include: [
        "children",
        "tooltipPosition",
        ...Object.keys(colorArgTypes),
        ...Object.keys(spacingArgTypes),
        ...Object.keys(borderArgTypes),
        ...Object.keys(typographyArgTypes),
      ],
    },
  },
  decorators: [
    (Story) => (
      <OakBox $ma="space-between-xxl">
        <Story />
      </OakBox>
    ),
  ],
  args: {
    children: "Hello there",
    tooltipPosition: "bottom-left",
  },
};
export default meta;

type Story = StoryObj<typeof InternalTooltip>;

export const Default: Story = {
  render: (args) => <InternalTooltip {...args} />,
};

export const WithStyling: Story = {
  render: (args) => (
    <InternalTooltip
      {...args}
      $background="lemon50"
      $ph="inner-padding-xl"
      $pv="inner-padding-m"
      $color="text-primary"
      $borderRadius="border-radius-m"
      $font="heading-light-7"
      $mb="space-between-xs"
    />
  ),
};
