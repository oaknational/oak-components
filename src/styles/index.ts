export * from "@/styles/theme/color";
