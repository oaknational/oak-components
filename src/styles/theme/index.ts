export { oakColorTokens, oakUiRoleTokens } from "./color";

export type {
  OakColorToken,
  OakCombinedColorToken,
  OakUiRoleToken,
  OakColorFilterToken,
} from "./color";

export {
  oakAllSpacingTokens,
  oakInnerPaddingTokens,
  oakSpaceBetweenTokens,
} from "./spacing";

export type {
  OakAllSpacingToken,
  OakCombinedSpacingToken,
  OakInnerPaddingToken,
  OakSpaceBetweenToken,
} from "./spacing";

export { oakDropShadowTokens } from "./dropShadow";
export type { OakDropShadowToken } from "./dropShadow";

export { oakOpacityTokens } from "./opacity";
export type { OakOpacityToken } from "./opacity";

export { oakTransitionTokens } from "./transitions";
export type { OakTransitionToken } from "./transitions";

export { oakBorderRadiusTokens, oakBorderWidthTokens } from "./borders";
export type { OakBorderRadiusToken, OakBorderWidthToken } from "./borders";

export type { OakTheme } from "./theme";

export { oakFontTokens, oakFontSizeTokens } from "./typography";
export type { OakFontToken, OakFontSizeToken } from "./typography";

export { oakZIndexTokens } from "./zIndex";

export { oakDefaultTheme } from "./default.theme";
