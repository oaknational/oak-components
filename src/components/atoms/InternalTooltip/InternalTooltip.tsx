import React, { ReactNode } from "react";
import styled, { css } from "styled-components";

import { OakFlex, OakFlexProps } from "../OakFlex";

import { OakCombinedColorToken } from "@/styles";
import { parseColor } from "@/styles/helpers/parseColor";
import { ColorStyleProps } from "@/styles/utils/colorStyle";
import { responsiveStyle } from "@/styles/utils/responsiveStyle";
import { parseSpacing } from "@/styles/helpers/parseSpacing";

export type InternalTooltipProps = OakFlexProps & {
  children?: ReactNode;
  tooltipPosition?: "bottom-left" | "bottom-right" | "top-left" | "top-right";
  id?: string;
};

const StyledFlex = styled(OakFlex)`
  width: max-content;
  pointer-events: none;
`;

type StyledSvgProps = {
  $fill: ColorStyleProps["$background"];
  $tooltipPosition?: InternalTooltipProps["tooltipPosition"];
};

const ARROW_SIZE = parseSpacing("all-spacing-4");

const StyledSvg = styled.svg<StyledSvgProps>`
  position: absolute;
  ${({ $tooltipPosition }) => {
    switch ($tooltipPosition) {
      case "bottom-right":
        return css`
          top: -${ARROW_SIZE};
          right: ${parseSpacing("all-spacing-0")};
          transform: scale(-1, -1);
        `;
      case "top-right":
        return css`
          bottom: -${ARROW_SIZE};
          right: ${parseSpacing("all-spacing-0")};
          transform: scaleX(-1);
        `;
      case "top-left":
        return css`
          bottom: -${ARROW_SIZE};
          left: ${parseSpacing("all-spacing-0")};
        `;
      default:
        return css`
          top: -${ARROW_SIZE};
          left: ${parseSpacing("all-spacing-0")};
          transform: scaleY(-1);
        `;
    }
  }}
  ${responsiveStyle<StyledSvgProps, OakCombinedColorToken>(
    "fill",
    (props) => props.$fill,
    parseColor,
  )}
`;

/**
 * A primitive tooltip to be used as a basis for more opinionated UI components.
 */
export const InternalTooltip = ({
  children,
  $background = "black",
  $color = "text-inverted",
  tooltipPosition = "bottom-left",
  id,
  ...props
}: InternalTooltipProps) => {
  return (
    <StyledFlex
      role="tooltip"
      id={id}
      data-rac
      {...props}
      $position="relative"
      $background={$background}
      $color={$color}
      $maxWidth={["all-spacing-20", "all-spacing-22"]}
    >
      {children}
      <StyledSvg
        width={ARROW_SIZE}
        height={ARROW_SIZE}
        $fill={$background}
        $tooltipPosition={tooltipPosition}
        data-testid="tooltip-arrow"
      >
        <path d="M0 0H16L8 8L0 16V0Z" />
      </StyledSvg>
    </StyledFlex>
  );
};
