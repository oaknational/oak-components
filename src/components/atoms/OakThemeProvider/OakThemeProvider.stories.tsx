import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakThemeProvider } from "./OakThemeProvider";

import { OakFlex } from "@/components/atoms/OakFlex";
import { oakDefaultTheme } from "@/styles";

const meta: Meta<typeof OakThemeProvider> = {
  component: OakThemeProvider,
  tags: ["autodocs"],
  title: "components/atoms/OakThemeProvider",
  argTypes: {},
  parameters: {
    controls: {
      include: [],
      sort: "none",
    },
  },
};
export default meta;

type Story = StoryObj<typeof OakThemeProvider>;

export const DefaultOakThemeProvider: Story = () => (
  <OakThemeProvider theme={oakDefaultTheme}>
    <OakFlex
      $background="bg-decorative1-main"
      $color="text-inverted"
      $ba="border-solid-m"
      $pa="spacing-24"
      $borderRadius="border-radius-l"
      $borderColor="border-primary"
    >
      Get a head-start on your lesson planning using quality-checked resources
      you can download and adapt for free.
    </OakFlex>
  </OakThemeProvider>
);

DefaultOakThemeProvider.args = {};
