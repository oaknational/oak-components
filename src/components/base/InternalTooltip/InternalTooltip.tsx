import React, { ReactNode } from "react";
import styled from "styled-components";

import { OakFlex, OakFlexProps } from "../OakFlex";

import { OakCombinedColorToken } from "@/styles";
import { parseColor } from "@/styles/helpers/parseColor";
import { ColorStyleProps } from "@/styles/utils/colorStyle";
import { responsiveStyle } from "@/styles/utils/responsiveStyle";

export type InternalTooltipProps = OakFlexProps & {
  isOpen: boolean;
  tooltip: ReactNode;
  children?: ReactNode;
};

const StyledFlex = styled(OakFlex)`
  width: max-content;
  pointer-events: none;
`;

type StyledSvgProps = {
  $fill: ColorStyleProps["$background"];
};

const StyledSvg = styled.svg<StyledSvgProps>`
  position: absolute;
  bottom: -16px;
  left: 0;
  ${responsiveStyle<StyledSvgProps, OakCombinedColorToken>(
    "fill",
    (props) => props.$fill,
    parseColor,
  )}
`;

export const InternalTooltip = ({
  isOpen,
  children,
  tooltip,
  $background = "black",
  $font = "heading-light-7",
  $color = "text-inverted",
  $borderRadius,
  ...props
}: InternalTooltipProps) => {
  return (
    <OakFlex $position="relative" $width="fit-content" $height="fit-content">
      {isOpen && (
        <OakFlex
          role="tooltip"
          $position="absolute"
          $top="all-spacing-0"
          $left="all-spacing-0"
          $transform="translateY(calc(-100% - 16px))"
          $zIndex="modal-dialog"
          $flexDirection="column"
        >
          <StyledFlex
            {...props}
            $position="relative"
            $background={$background}
            $color={$color}
            $font={$font}
            $btr={$borderRadius}
            $bbrr={$borderRadius}
            $maxWidth={["all-spacing-20", "all-spacing-22"]}
          >
            {tooltip}
            <StyledSvg width="16" height="16" $fill={$background}>
              <path d="M0 0H16L8 8L0 16V0Z" />
            </StyledSvg>
          </StyledFlex>
        </OakFlex>
      )}
      {children}
    </OakFlex>
  );
};
