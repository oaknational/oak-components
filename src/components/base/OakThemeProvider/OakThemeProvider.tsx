import React from "react";
import { ThemeProvider } from "styled-components";

import { OakTheme } from "@/styles";

export type OakThemeProviderProps = {
  theme: OakTheme;
  children: React.ReactNode;
};

export const OakThemeProvider = ({
  theme,
  children,
}: OakThemeProviderProps) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
};
