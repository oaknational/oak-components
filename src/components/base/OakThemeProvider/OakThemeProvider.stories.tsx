import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import { OakThemeProvider } from "./OakThemeProvider";

import { OakFlex } from "@/components/base/OakFlex";
import { oakDefaultTheme } from "@/styles";

/**
 *
 * OakThemeProvider wraps Styled Components ThemeProvider allowing the use of the custom type OakTheme.
 *
 */

const meta: Meta<typeof OakThemeProvider> = {
  component: OakThemeProvider,
  tags: ["autodocs"],
  title: "components/base/OakThemeProvider",
  argTypes: {},
  parameters: {
    controls: {
      include: [],
      sort: "none",
    },
  },
};
export default meta;

type Story = StoryObj<typeof OakThemeProvider>;

export const DefaultOakThemeProvider: Story = () => (
  <OakThemeProvider theme={oakDefaultTheme}>
    <OakFlex
      $background="bg-decorative1-main"
      $color="text-inverted"
      $ba="border-solid-m"
      $pa="inner-padding-xl"
      $borderRadius="border-radius-l"
      $borderColor="border-primary"
    >
      Get a head-start on your lesson planning using quality-checked resources
      you can download and adapt for free.
    </OakFlex>
  </OakThemeProvider>
);

DefaultOakThemeProvider.args = {};
