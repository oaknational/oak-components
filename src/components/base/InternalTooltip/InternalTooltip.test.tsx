import { create } from "react-test-renderer";
import React from "react";
import "@testing-library/jest-dom";

import { OakThemeProvider } from "../OakThemeProvider";

import { InternalTooltip } from "./InternalTooltip";

import renderWithTheme from "@/test-helpers/renderWithTheme";
import { oakDefaultTheme } from "@/styles";

describe(InternalTooltip, () => {
  it("matches snapshot", () => {
    const tree = create(
      <OakThemeProvider theme={oakDefaultTheme}>
        <InternalTooltip tooltip="Hello there" isOpen>
          Trigger!
        </InternalTooltip>
      </OakThemeProvider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders children", () => {
    const { getByText } = renderWithTheme(
      <InternalTooltip tooltip="Hello there" isOpen={false}>
        Trigger!
      </InternalTooltip>,
    );

    expect(getByText("Trigger!")).toBeInTheDocument();
  });

  it("renders a tooltip when `isOpen` is true", () => {
    const { rerender, queryByRole } = renderWithTheme(
      <InternalTooltip tooltip="Hello there" isOpen={false} />,
    );

    expect(queryByRole("tooltip")).not.toBeInTheDocument();

    rerender(
      <InternalTooltip tooltip="Hello there" isOpen>
        Trigger!
      </InternalTooltip>,
    );

    expect(queryByRole("tooltip")).toBeInTheDocument();
  });
});
