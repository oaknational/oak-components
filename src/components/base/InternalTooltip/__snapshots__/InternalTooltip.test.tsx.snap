// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InternalTooltip matches snapshot 1`] = `
.c0 {
  position: relative;
  width: -webkit-fit-content;
  width: -moz-fit-content;
  width: fit-content;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  font-family: Lexend,sans-serif;
}

.c2 {
  position: absolute;
  top: 0rem;
  left: 0rem;
  -webkit-transform: translateY(calc(-100% - 16px));
  -ms-transform: translateY(calc(-100% - 16px));
  transform: translateY(calc(-100% - 16px));
  font-family: Lexend,sans-serif;
  z-index: 300;
}

.c4 {
  position: relative;
  max-width: 22.5rem;
  color: #ffffff;
  background: #222222;
  font-family: Lexend,sans-serif;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c5 {
  width: -webkit-max-content;
  width: -moz-max-content;
  width: max-content;
  pointer-events: none;
}

.c6 {
  position: absolute;
  bottom: -16px;
  left: 0;
  fill: #222222;
}

@media (min-width:750px) {
  .c4 {
    max-width: 40rem;
  }
}

<div
  className="c0 c1"
>
  <div
    className="c2 c3"
    role="tooltip"
  >
    <div
      className="c4 c1 c5"
    >
      Hello there
      <svg
        className="c6"
        height="16"
        width="16"
      >
        <path
          d="M0 0H16L8 8L0 16V0Z"
        />
      </svg>
    </div>
  </div>
  Trigger!
</div>
`;
